src/main/java/com/encora/todoapp_be/service/TodoService.java

```
@@ -67,6 +69,46 @@ public TodoModel addTodo(TodoModel todo) {
         return todo;
     }
 
+    public TodoModel updateTodo(UpdateTodoDTO updateTodo) {
+        for (int i = 0; i < todos.size(); i++) {
+            if (todos.get(i).getId().equals(updateTodo.getId())) {
+                if(updateTodo.getText() != null) {
+                    todos.get(i).setText(updateTodo.getText());
+                }
+                if(updateTodo.getDueDate() != null) {
+                    todos.get(i).setDueDate(updateTodo.getDueDate());
+                }
+                if(updateTodo.getPriority() != null) {
+                    todos.get(i).setPriority(updateTodo.getPriority());
+                }
+                return todos.get(i);
+            }
+        }
+        return null; // or throw an exception if not found
+    }
+
+    public TodoModel markTodoAsDone(Long id) {
+        for (TodoModel todo : todos) {
+            if (todo.getId().equals(id) && todo.isCompleted() == false) {
+                todo.setCompleted(true);
+                todo.setDoneDate();
+                return todo;
+            }
+        }
+        return null;
```

why you are returning null here if you not find the todo you could throw an error